nmt.go

```diff
@@ -0,0 +1,242 @@
+// generate any proof and we return an empty range (0,0) to
+// indicate that this namespace is not contained in the tree.
+func (n NamespacedMerkleTree) ProveNamespace(nID namespace.ID) (Proof, error) {
+	// In the cases (nID < minNID) or (maxNID < nID),
+	// return empty range and no proof:
+	if nID.Less(n.minNID) || n.maxNID.Less(nID) {
+		return NewEmptyRangeProof(), nil
+	}
+
+	found, proofStart, proofEnd := n.foundInRange(nID)
+	if !found {
+		// To generate a proof for an absence we calculate the
+		// range the namespace would be in:
+		proofStart, proofEnd = n.calculateAbsenceRange(nID)
+	}
+	// At this point we either found the namespace in the tree or calculated
+	// the range it would be in (to generate a proof of absence and to return
+	// the corresponding leaf hashes).
+	subTreeHasher := internal.NewCachedSubtreeHasher(n.leafHashes, n.treeHasher)
+	var err error
+	proof, err := merkletree.BuildRangeProof(proofStart, proofEnd, subTreeHasher)
+	if err != nil {
+		// This should never happen.
+		// TODO would be good to back this by more tests and fuzzing.
+		return Proof{}, fmt.Errorf(
+			"unexpected err: %w on nID: %v, range: [%v, %v)",
+			err,
+			nID,
+			proofStart,
+			proofEnd,
+		)
+	}
+
+	if found {
+		return NewInclusionProof(proofStart, proofEnd, proof), nil
+	}
+	return NewAbsenceProof(proofStart, proofEnd, proof, n.leafHashes[proofStart:proofEnd]), nil
+}
+
+// Get returns leaves for the given namespace.ID.
+func (n NamespacedMerkleTree) Get(nID namespace.ID) []namespace.PrefixedData {
+	_, start, end := n.foundInRange(nID)
+	return n.leafs[start:end]
+}
+
+// Get is a convenience method returns leaves for the given namespace.ID
+// together with the proof for that namespace. It returns the same result
+// as calling the combination of Get(nid) and ProveNamespace(nid).
+func (n NamespacedMerkleTree) GetWithProof(nID namespace.ID) ([]namespace.PrefixedData, Proof, error) {
+	data := n.Get(nID)
+	proof, err := n.ProveNamespace(nID)
+	return data, proof, err
+}
+
+func (n NamespacedMerkleTree) calculateAbsenceRange(nID namespace.ID) (int, int) {
+	foundRangeStart := false
+	proofStart, proofEnd := 0, 0
+	var prevLeaf namespace.PrefixedData
+	for index, curLeaf := range n.leafs {
+		if index == 0 {
+			prevLeaf = curLeaf
+			continue
+		}
+		prevNs := prevLeaf.NamespaceID()
+		currentNs := curLeaf.NamespaceID()
+		// Note that here we would also care for the case
+		// current < nId < prevNs
+		// but we only allow pushing leafs with ascending namespaces;
+		// i.e. prevNs <= currentNs is always true.
+		// Also we only check for strictly smaller: prev < nid < current
+		// because if we either side was equal, we would have found the
+		// namespace before.
+		if prevNs.Less(nID) && nID.Less(currentNs) {
+			if !foundRangeStart {
+				foundRangeStart = true
+				proofStart = index
+			}
+			proofEnd = index + 1
+		}
+		prevLeaf = curLeaf
+	}
+	return proofStart, proofEnd
```

Comment:
Doesn't the lexicographic sorting imply that `proofStart == proofEnd-1` is always true? 

Meaning that a proof of absence will always be just *one* leaf hash: namely at the single index i for which `n_{i-1} < nID < n_i`; if that was true for another index `j`:  `n_{j-1} < nID < n_j` that would imply for instance n_i < n_j but also n_j < n_i (which is impossible and hence proves that there can't be another index). 

Anything wrong with the reasoning here?

Reply:
Yes, this is the case in the prototype: https://github.com/lazyledger/lazyledger-prototype/blob/master/simpleblock.go#L133

Reply:
That's not obvious: in the code you share `proofStart` gets set to `index` once but `proofEnd` could be set to `index + 1` multiple times (or at least it is not obvious that it only gets set once), and index is increasing (hence proofEnd could be anything > proofStart). I think it will only be set once if the data is pushed sorted by namespace (which is enforced here); so we can safely assume one leaf.